PhysicsTools/TensorFlow/interface/Tensor.h

```diff
@@ -0,0 +1,450 @@
+    }
+
+    // sets the values of a rank 3 tensor along an axis for a fixed position on the other axes
+    template <typename T>
+    inline void setVector(int axis, Shape a, Shape b, std::vector<T>& v)
+    {
+        assertRank(3);
+        Shape pos[2] = { a, b };
+        setVectorAtPos<T>(axis, pos, v);
+    }
+
+    // sets the values of a rank 4 tensor along an axis for a fixed position on the other axes
+    template <typename T>
+    inline void setVector(int axis, Shape a, Shape b, Shape c, std::vector<T>& v)
+    {
+        assertRank(4);
+        Shape pos[3] = { a, b, c };
+        setVectorAtPos<T>(axis, pos, v);
+    }
+
+    // sets n elements to a constant value starting from a certain position, negative n means all
+    template <typename T>
+    void fillValuesAtPos(T v, int n, Shape* pos);
+
+    // sets n elements of a rank 1 tensor to a constant value starting from a certain position,
+    // negative n means all
+    template <typename T>
+    inline void fillValues(T v, int n, Shape i)
+    {
+        assertRank(1);
+        Shape pos[1] = { i };
+        fillValuesAtPos<T>(v, n, pos);
+    }
+
+    // sets n elements of a rank 2 tensor to a constant value starting from a certain position,
+    // negative n means all
+    template <typename T>
+    inline void fillValues(T v, int n, Shape i, Shape j)
+    {
+        assertRank(2);
+        Shape pos[2] = { i, j };
+        fillValuesAtPos<T>(v, n, pos);
+    }
+
+    // sets n elements of a rank 3 tensor to a constant value starting from a certain position,
+    // negative n means all
+    template <typename T>
+    inline void fillValues(T v, int n, Shape i, Shape j, Shape k)
+    {
+        assertRank(3);
+        Shape pos[3] = { i, j, k };
+        fillValuesAtPos<T>(v, n, pos);
+    }
+
+    // sets n elements of a rank 4 tensor to a constant value starting from a certain position,
+    // negative n means all
+    template <typename T>
+    inline void fillValues(T v, int n, Shape i, Shape j, Shape k, Shape l)
+    {
+        assertRank(4);
+        Shape pos[4] = { i, j, k, l };
+        fillValuesAtPos<T>(v, n, pos);
+    }
+
+private:
+    // the internal tensorflow tensor object
+    TF_Tensor* tf_tensor;
+
+    // array of cumulative shape products to accelerate indexing
+    int64_t* prod;
+
+    // compares a passed rank to the current rank and throws an exception when they do not match
+    inline void assertRank(int rank) const
+    {
+        if (getRank() != rank)
+        {
+            throw std::runtime_error("invalid rank to perform operation: "
+                + std::to_string(getRank()) + " (expected " + std::to_string(rank) + ")");
+        }
+    }
+};
+
+template <typename T>
+void Tensor::getPtrVectorAtPos(int axis, Shape* pos, std::vector<T*>& v)
+{
+    const int rank = getRank();
+
+    // special treatment of scalars
+    if (rank == 0)
+    {
+        throw std::runtime_error("vectors cannot be extracted from scalars");
```

Comment:
please change to edm::Exception
@Dr15Jones please remind me if there is a preference